# Basic Regex Homework

In [ ]:
# use print only as a function
# ignore this step as we are working in Python 2.7
#from __future__ import print_function

## Homework 1: FAA tower closures

A list of FAA tower closures has been copied from a [PDF](http://www.faa.gov/news/media/fct_closed.pdf) into the file **`faa.txt`**, which is stored in the **`data`** directory of the course repository.

In [1]:
# read the file into a single string
with open('../data/faa.txt') as f:
    data = f.read()

In [35]:
# check number of lines
g = open('../data/faa.txt').readlines()

In [37]:
print len(g)
print g[0:20]

173
['FAA Contract Tower Closure List\n', '(149 FCTs)\n', '3-22-2013\n', 'LOC\n', 'ID Facility Name City State\n', 'DHN DOTHAN RGNL DOTHAN AL\n', 'TCL TUSCALOOSA RGNL TUSCALOOSA AL\n', 'FYV DRAKE FIELD FAYETTEVILLE AR\n', 'TXK TEXARKANA RGNL-WEBB FIELD TEXARKANA AR\n', 'GEU GLENDALE MUNI GLENDALE AZ\n', 'GYR PHOENIX GOODYEAR GOODYEAR AZ\n', 'IFP LAUGHLIN/BULLHEAD INTL BULLHEAD CITY AZ\n', 'RYN RYAN FIELD TUCSON AZ\n', 'FUL FULLERTON MUNI FULLERTON CA\n', 'MER CASTLE ATWATER CA\n', 'OXR OXNARD OXNARD CA\n', 'RAL RIVERSIDE MUNI RIVERSIDE CA\n', 'RNM RAMONA RAMONA CA\n', 'SAC SACRAMENTO EXECUTIVE SACRAMENTO CA\n', 'SDM BROWN FIELD MUNI SAN DIEGO CA\n']


In [2]:
# check the number of characters
len(data)

5573

In [3]:
# examine the first 500 characters
print(data[0:500])

FAA Contract Tower Closure List
(149 FCTs)
3-22-2013
LOC
ID Facility Name City State
DHN DOTHAN RGNL DOTHAN AL
TCL TUSCALOOSA RGNL TUSCALOOSA AL
FYV DRAKE FIELD FAYETTEVILLE AR
TXK TEXARKANA RGNL-WEBB FIELD TEXARKANA AR
GEU GLENDALE MUNI GLENDALE AZ
GYR PHOENIX GOODYEAR GOODYEAR AZ
IFP LAUGHLIN/BULLHEAD INTL BULLHEAD CITY AZ
RYN RYAN FIELD TUCSON AZ
FUL FULLERTON MUNI FULLERTON CA
MER CASTLE ATWATER CA
OXR OXNARD OXNARD CA
RAL RIVERSIDE MUNI RIVERSIDE CA
RNM RAMONA RAMONA CA
SAC SACRAMENTO EXECU


In [8]:
# examine the last 500 characters
print(data[-550:])

TACOMA WA
YKM YAKIMA AIR TERMINAL/ MCALLISTER FIELD YAKIMA WA
CWA CENTRAL WISCONSIN MOSINEE WI
EAU CHIPPEWA VALLEY RGNL EAU CLAIRE WI
ENW KENOSHA RGNL KENOSHA WI
Page 3 of 4
FAA Contract Tower Closure List
(149 FCTs)
3-22-2013
LOC
ID Facility Name City State
JVL SOUTHERN WISCONSIN RGNL JANESVILLE WI
LSE LA CROSSE MUNI LA CROSSE WI
MWC LAWRENCE J TIMMERMAN MILWAUKEE WI
OSH WITTMAN RGNL OSHKOSH WI
UES WAUKESHA COUNTY WAUKESHA WI
HLG WHEELING OHIO CO WHEELING WV
LWB GREENBRIER VALLEY LEWISBURG WV
PKB MID-OHIO VALLEY RGNL PARKERSBURG WV
Page 4 of 4


Your assignment is to **create a list of tuples** containing the **tower IDs** and the **states** they are located in.

Here is the **expected output:**

> `faa = [('DHN', 'AL'), ('TCL', 'AL'), ..., ('PKB', 'WV')]`

In [25]:
# import the regex module
import re

In [55]:
# answer to part 1
faa = re.findall(r'([A-Z]{3}) .+ ([A-Z]{2})\n', data[1:])

In [66]:
# check the len
faa_length = len(faa)

In [67]:
#inspect first 20 tuples
print faa[0:20]

[('DHN', 'AL'), ('TCL', 'AL'), ('FYV', 'AR'), ('TXK', 'AR'), ('GEU', 'AZ'), ('GYR', 'AZ'), ('IFP', 'AZ'), ('RYN', 'AZ'), ('FUL', 'CA'), ('MER', 'CA'), ('OXR', 'CA'), ('RAL', 'CA'), ('RNM', 'CA'), ('SAC', 'CA'), ('SDM', 'CA'), ('SNS', 'CA'), ('VCV', 'CA'), ('WHP', 'CA'), ('WJF', 'CA'), ('BDR', 'CT')]


In [68]:
# inspect last 20 tuples
print faa[-20:]

[('VCT', 'TX'), ('OGD', 'UT'), ('PVU', 'UT'), ('LYH', 'VA'), ('OLM', 'WA'), ('RNT', 'WA'), ('SFF', 'WA'), ('TIW', 'WA'), ('YKM', 'WA'), ('CWA', 'WI'), ('EAU', 'WI'), ('ENW', 'WI'), ('JVL', 'WI'), ('LSE', 'WI'), ('MWC', 'WI'), ('OSH', 'WI'), ('UES', 'WI'), ('HLG', 'WV'), ('LWB', 'WV'), ('PKB', 'WV')]


As a **bonus task**, use regular expressions to extract the **number of closures** listed in the second line of the file (149), and then use an **assertion** to check that the number of closures is equal to the length of the `faa` list.

In [74]:
# extract the number of closures only
num_closures = re.search(r'[0-9]{3}', data)
num_closures = int(num_closures.group())
num_closures == faa_length

True

In [79]:
# create a function to use an assertion to check that the number of closures equals the lengh of faa list
def num_check(number):
    assert (number == len(faa)), "the length of the faa doc and the number of closures don't match!"
    return (number)

In [81]:
# check the assertion
num_check(num_closures)

149

## Homework 2: Stack Overflow reputation

I have downloaded my **Stack Overflow reputation history** into the file **`reputation.txt`**, which is stored in the **`data`** directory of the course repository. (If you are a Stack Overflow user with a reputation of 10 or more, you should be able to [download your own reputation history](http://stackoverflow.com/reputation).)

We are only interested in the lines that **begin with two dashes**, such as:

> `-- 2012-08-30 rep +5    = 6`

That line can be interpreted as follows: "On 2012-08-30, my reputation increased by 5, bringing my reputation total to 6."

Your assignment is to **create a list of tuples** containing only these dated entries, including the **date**, **reputation change** (regardless of whether it is positive/negative/zero), and **running total**.

Here is the **expected output:**

> `rep = [('2012-08-30', '+5', '6'), ('2012-12-11', '+10', '16'), ...,  ('2015-10-14', '-1', '317')]`

In [1]:
# load the data into the ipynb
with open("../data/reputation.txt") as f:
    data = f.read()

In [2]:
# examine the first 500 characters
print data[0:500]

total votes: 36
 2  12201376 (5)
-- 2012-08-30 rep +5    = 6         
 2  13822612 (10)
-- 2012-12-11 rep +10   = 16        
 2  13822612 (10)
-- 2013-03-20 rep +10   = 26        
-- 2013-12-05 rep 0     = 26        
-- 2014-01-25 rep 0     = 26        
 16  7141669 (2)
-- 2014-03-19 rep +2    = 28        
 1  12202249 (2)
-- 2014-05-11 rep +2    = 30        
 16 23599806 (2)
 2  23597220 (10)
-- 2014-05-12 rep +12   = 42        
 2  13822612 (10)
-- 2014-06-12 rep +10   = 52        
 2  2359722


In [3]:
# examine the last 500 characters
print data[-500:]

5-09-07 rep +10   = 318       
 3   1839257 (-1)
-- 2015-10-14 rep -1    = 317       

** rep today: 0
** rep this week (2015-11-08 - 2015-11-14): 0
** rep this month (2015-11-01 - 2015-11-30): 0
** rep this quarter (2015-10-01 - 2015-12-31): -1
** rep this year (2015-01-01 - 2015-12-31): 109
** rep from bonuses: 100
** total rep 317 :)

days represented 34
rep cap was reached via rep from upvotes *only* on 0 days
earned at least 200 reputation on 0 days
earned 10 reputation from suggested edits


In [24]:
# import the regex module
# imprt the pandas module
import re
import pandas as pd

In [5]:
# Homework Basic Regex 2
rep = re.findall(r'[-]{2} ([\d]{4}-[\d]{2}-[\d]{2}) [a-z]{3} (.[\d]).+ = ([\d]+)', data)
rep

[('2012-08-30', '+5', '6'),
 ('2012-12-11', '+1', '16'),
 ('2013-03-20', '+1', '26'),
 ('2014-03-19', '+2', '28'),
 ('2014-05-11', '+2', '30'),
 ('2014-05-12', '+1', '42'),
 ('2014-06-12', '+1', '52'),
 ('2014-06-26', '+1', '62'),
 ('2014-09-03', '+1', '72'),
 ('2014-11-14', '+1', '82'),
 ('2014-11-18', '+2', '84'),
 ('2014-12-08', '+2', '86'),
 ('2014-12-09', '+1', '96'),
 ('2014-12-12', '+2', '98'),
 ('2014-12-24', '+1', '108'),
 ('2015-02-20', '+1', '118'),
 ('2015-03-28', '+1', '128'),
 ('2015-04-26', '+1', '138'),
 ('2015-05-05', '+1', '148'),
 ('2015-05-26', '+1', '158'),
 ('2015-05-27', '+2', '178'),
 ('2015-07-03', '+1', '188'),
 ('2015-07-22', '+1', '298'),
 ('2015-08-21', '+1', '308'),
 ('2015-09-07', '+1', '318'),
 ('2015-10-14', '-1', '317')]

As a **bonus task**, convert this list of tuples into a **pandas DataFrame**. It should have appropriate column names, and the second and third columns should be of type integer (rather than string/object).

In [13]:
# import the pandas module
dates = []
increment = []
score = []
for each in rep:
    dates.append(each[0])
    increment.append(int(each[1]))
    score.append(int(each[2]))

In [20]:
# check to make sure the conversion worked from string to int
# inspect the first few elements of each newly created list
print dates[0:5]
print increment[-5:]
print score[-5:]
print type(increment[-1]) ; type(score[0])

['2012-08-30', '2012-12-11', '2013-03-20', '2014-03-19', '2014-05-11']
[1, 1, 1, 1, -1]
[188, 298, 308, 318, 317]
<type 'int'>


int

In [27]:
# create the dictionary to convert to pd DF
d = {'a' : dates, 'b': increment, 'c': score}

In [32]:
# create the pd DataFrame
bonus = pd.DataFrame(d)
bonus.columns = ['Date', 'Increment', 'Total_Score']

In [33]:
# inspect visually to make sure worked as intended.
bonus

Date  Increment  Total_Score
0   2012-08-30          5            6
1   2012-12-11          1           16
2   2013-03-20          1           26
3   2014-03-19          2           28
4   2014-05-11          2           30
5   2014-05-12          1           42
6   2014-06-12          1           52
7   2014-06-26          1           62
8   2014-09-03          1           72
9   2014-11-14          1           82
10  2014-11-18          2           84
11  2014-12-08          2           86
12  2014-12-09          1           96
13  2014-12-12          2           98
14  2014-12-24          1          108
15  2015-02-20          1          118
16  2015-03-28          1          128
17  2015-04-26          1          138
18  2015-05-05          1          148
19  2015-05-26          1          158
20  2015-05-27          2          178
21  2015-07-03          1          188
22  2015-07-22          1          298
23  2015-08-21          1          308
24  2015-09-07          1          318
25  2015-10-14         -1          317

In [37]:
# Alternative option to create column names initially
test_d = {'Date':dates, 'Point_Change': increment, 'Total_Score': score}

In [38]:
# create new test df with col names already created
test = pd.DataFrame(test_d)

In [47]:
test

Date  Point_Change  Total_Score
0   2012-08-30             5            6
1   2012-12-11             1           16
2   2013-03-20             1           26
3   2014-03-19             2           28
4   2014-05-11             2           30
5   2014-05-12             1           42
6   2014-06-12             1           52
7   2014-06-26             1           62
8   2014-09-03             1           72
9   2014-11-14             1           82
10  2014-11-18             2           84
11  2014-12-08             2           86
12  2014-12-09             1           96
13  2014-12-12             2           98
14  2014-12-24             1          108
15  2015-02-20             1          118
16  2015-03-28             1          128
17  2015-04-26             1          138
18  2015-05-05             1          148
19  2015-05-26             1          158
20  2015-05-27             2          178
21  2015-07-03             1          188
22  2015-07-22             1          298
23  2015-08-21             1          308
24  2015-09-07             1          318
25  2015-10-14            -1          317

In [43]:
# create new Df using column slicing to reorder
# testing to see if this works because the first dictionary I created for the assignment was re-ordered alphabetically by Pandas 
# and put the columns in an order that I didn't want.  
new_test = test[['Total_Score', 'Date', 'Point_Change']]

In [45]:
# Alternative option to re-order feature columns by referecning the column index
slice2 = test[[2,0,1]]

In [46]:
# inspect visually to make sure it worked
slice2

Total_Score        Date  Point_Change
0             6  2012-08-30             5
1            16  2012-12-11             1
2            26  2013-03-20             1
3            28  2014-03-19             2
4            30  2014-05-11             2
5            42  2014-05-12             1
6            52  2014-06-12             1
7            62  2014-06-26             1
8            72  2014-09-03             1
9            82  2014-11-14             1
10           84  2014-11-18             2
11           86  2014-12-08             2
12           96  2014-12-09             1
13           98  2014-12-12             2
14          108  2014-12-24             1
15          118  2015-02-20             1
16          128  2015-03-28             1
17          138  2015-04-26             1
18          148  2015-05-05             1
19          158  2015-05-26             1
20          178  2015-05-27             2
21          188  2015-07-03             1
22          298  2015-07-22             1
23          308  2015-08-21             1
24          318  2015-09-07             1
25          317  2015-10-14            -1